clients/wallet/src/resubmissions.rs

```diff
@@ -0,0 +1,795 @@
+		}
+	}
+
+	/// Returns:
+	/// * `TransactionResponse` for successful resubmission;
+	/// * None for errors that cannot be resubmitted;
+	/// * An error that can be resubmitted again
+	///
+	/// This function determines whether an error is up for resubmission or not:
+	/// `tx_bad_seq` or `SequenceNumberAlreadyUsed` can be resubmitted by updating the sequence
+	/// number `tx_internal_error` should be resubmitted again
+	/// other errors must be logged and removed from cache.
+	async fn handle_error(&self, error: Error) -> Result<Option<TransactionResponse>, Error> {
+		match &error {
+			Error::HorizonSubmissionError { reason, envelope_xdr, .. } => match &reason[..] {
+				"tx_bad_seq" =>
+					return self.handle_tx_bad_seq_error_with_xdr(envelope_xdr).await.map(Some),
+				"tx_internal_error" =>
+					return self.handle_tx_internal_error(envelope_xdr).await.map(Some),
+				_ => {
+					if let Ok(env) = decode_to_envelope(envelope_xdr) {
+						if let Some(sequence) = env.sequence_number() {
+							self.remove_tx_envelope_from_cache(sequence);
+						}
+					};
+
+					tracing::error!(
+						"handle_error(): Unrecoverable HorizonSubmissionError: {error:?}"
+					);
+				},
+			},
+			Error::CacheError(CacheError {
+				kind: CacheErrorKind::SequenceNumberAlreadyUsed,
+				envelope,
+				..
+			}) => {
+				if let Some(transaction_envelope) = envelope {
+					return self
+						.handle_tx_bad_seq_error_with_envelope(transaction_envelope.clone())
+						.await
+						.map(Some)
+				}
+
+				tracing::warn!("handle_error(): SequenceNumberAlreadyUsed error but no envelope");
+			},
+			_ => tracing::warn!("handle_error(): Unrecoverable error in Stellar wallet: {error:?}"),
+		}
+
+		// the error found is not recoverable, and cannot be resubmitted again.
+		Ok(None)
+	}
+
+	async fn handle_tx_internal_error(
+		&self,
+		envelope_xdr_as_str_opt: &Option<String>,
+	) -> Result<TransactionResponse, Error> {
+		let mut envelope = decode_to_envelope(envelope_xdr_as_str_opt)?;
+		self.sign_envelope(&mut envelope)?;
+
+		self.submit_transaction(envelope).await
+	}
+}
+
+// handle tx_bad_seq
+#[cfg_attr(test, mockable)]
+impl StellarWallet {
+	async fn handle_tx_bad_seq_error_with_xdr(
+		&self,
+		envelope_xdr_as_str_opt: &Option<String>,
+	) -> Result<TransactionResponse, Error> {
+		let tx_envelope = decode_to_envelope(envelope_xdr_as_str_opt)?;
+		self.handle_tx_bad_seq_error_with_envelope(tx_envelope).await
+	}
+
+	async fn handle_tx_bad_seq_error_with_envelope(
+		&self,
+		tx_envelope: TransactionEnvelope,
+	) -> Result<TransactionResponse, Error> {
+		let tx = tx_envelope.get_transaction().ok_or(DecodeError)?;
+
+		// Remove original transaction.
+		// The same envelope will be saved again using a different sequence number
+		self.remove_tx_envelope_from_cache(tx.seq_num);
+
+		// Check if we already submitted this transaction
+		if !self.is_transaction_already_submitted(&tx).await {
+			return self.bump_sequence_number_and_submit(tx).await
+		}
```

Comment:
```suggestion
		// Check if we already submitted this transaction
		if !self.is_transaction_already_submitted(&tx).await {
			self.bump_sequence_number_and_submit(tx).await
			self.remove_tx_envelope_from_cache(tx.seq_num);
			return;
		}
```

Reply:
i called `remove_tx_envelope_from_cache()` first, before the `bump_sequence_number_and_submit()`